### DISCUSSION TITLE
zwlr_layer_shell_v1 supported

### DISCUSSION THREAD

**Discussion by Decodetalkers** [slint-ui/slint#1765] [Q&A]
This protocol is useful when on wayland, can make a float window on screen. Will you support it in slint in the feature release?

**Comment by ogoffart** [comment]
What is your use case?
Is that what slin's PopupWindow does? (This would be, I suppose, an implementation detail of such primitive.)

We would need support for that in the winit crate that is handling our wayland interaction.

**Reply by Decodetalkers** [reply]
> What is your use case? Is that what slin's PopupWindow does? (This would be, I suppose, an implementation detail of such primitive.)
> 
> We would need support for that in the winit crate that is handling our wayland interaction.

 https://github.com/l4l/yofi.git 
like this one
zwlr_layer_shell is useful, it will make a float window pined on the screen
You can view the details from https://wayland.app/protocols/wlr-layer-shell-unstable-v1
Crate example you can see in  https://github.com/Smithay/client-toolkit.git, the example of simple_layer.rs
Not only PopupWindow.. I need all windows can be set as layer.. At least be set in the center of the screen
If it can be a key in the slint , it will be better , I hope
and you had better test the example above on swaywm. Sway has full realize for the protocol

**Comment by ngortheone** [comment]
Layer Shell is a wayland protocol that allows an application to occupy an exclusive area on the screen (overlay, anchored to screen anchors etc.). This is highly useful for building shell components like bars, panels, login screens, etc.

Protocol description
https://wayland.app/protocols/wlr-layer-shell-unstable-v1
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/master/protocol/wlr-layer-shell-unstable-v1.xml

Demo of the protocol by Drew DeVault: https://www.youtube.com/watch?v=VuRXHJu5Kmg

Similar GTK project, that adds layer shell to gtk3
https://github.com/wmww/gtk-layer-shell

An example panel project that uses gtk-layer-shell
https://github.com/Alexays/Waybar

**Comment by tronical** [comment]
I think the concrete way of anchoring and layering surfaces is an implementation detail. As such, I'm not sure Slint can provide any functionality regarding the concrete protocol. However, Slint can and should provide access to the wayland display and the wayland surface for a given window, so that applications can connect that to their code to send/receive wlr specific messages to/from the compositor.

As such, I think we should provide access to the raw-window-handle's RawDisplayHandle and RawWindowHandle.

**Reply by ngortheone** [reply]
> Slint can and should provide access to the wayland display and the wayland surface for a given window

This should really be enough. Having access to the surface a programmer can use it to set layer shell role to it.

**Comment by Jaakkonen** [comment]
The earlier comments sound super good! Below are some of my thoughts as a first timer going through the code base:

I think this kind of hacks could work to get all the requirements for layer shell to one place
1. Make it possible to get mutable borrow of the platform, add `fn on_raw_window_handle_created(fn: Box<dyn Fn(Platform) -> ()>)` and `fn get_raw_display_handle() -> Option<PlatformDisplayHandle>` functions to `Platform`. However since both of these abstractions don't exist on all platforms (talking embedded here) I think that this would lead to some bloat.
2. Allow instantiation of platforms and add these `get_raw_x()` functions there. However this would make the setting of backends less consistent between apps and universal `SLINT_BACKEND` environment variable wouldn't work.
3. Essentially 1. but make platforms be public types that can be downcast_ref'd to. The special methods would be exported on per platform implementation level. IMO this wouldn't scale that well and people would end up just creating new requests for adding xyz bindings, the same problem applies to above approaches as well.
4. Expose initiation and creation of platforms to API users. Users would now be able to create custom platforms calling to existing implementations but adding functionality on top of those. However since the platforms definitely aren't thin wrappers and just getting to `winit` window creation for a newcomer to the project takes me quite some hassle (window_factory_fn goes quite deep but the main thing is that it returns `Rc<dyn WindowAdapter>` which for GLWindow implementation creates a winit window in `show` implementation of the sealed `WindowAdapterSealed` trait which then has the ability to get the raw pointers)

All of these options sound nice on the outside but there are quite many gotchas. Let's think about a case where someone wants to implement for example a shell component using zwlr_layer_shell_v1.
- The initialization control flow eventually goes to the platform creation function and then show method
- For `winit`s case the show method builds the window
  - Internally winit now connects to wayland socket, sends wl_display::get_registry, gets the registry, binds to wanted protocols
  - Now winit creates a wl_surface and creates xdg_surface from it. A window is shown on screen with random data until render function copies some pixels are put to it. (It may be wl_buffer that is just mmapped shared POSIX shm, or )
  - The control flow is now passed to Slint with a window just opened (or waiting for wayland commit, not sure about the details here)
  - Only now are the raw window and raw display handles available. These could be passed from the `fn show` to some Winit struct specific field.
  - Now the user can get the control flow, destroy the xdg_window, commit, create a layer shell, commit, associate the wl_surface to that layer shell, commit and have the thing originally wanted

There are some concerns about that, the first being that a window needs to be shown from Slints point of view first before any Wayland connections exist. For applications such as notification daemons, the window should only be shown when reacting to for example an IPC call. This window "flashing" could maybe be worked around by setting the window to be not visible in the winit builder (however their documentation seems to suggest that set_visible is not working for wayland)
Another ability that would be hard to implement this way is hiding the window for it to become shown later on. Wayland doesn't directly have any **hidden** surface property although all pixels can be set to 0 alpha (in which case mouse is still catched). GTK implements the hiding by destroying all wayland surfaces/shells/buffers and re-creating & copying image once setting visibility back to true.

Frankly the only way I see these nieche cases to be able to be addressed is to have the `Platform` open for outside implementations and having the ability to just take a renderer which gives rectangles of damaged areas (for software rendering) and writes the images to a provided buffer or to some EGL/DMABUF/Vulkan shared buffer. As winit doesn't implement the creation of these OpenGL/Vulkan/etc shared buffers I expect those to be handled somewhere in Slint (or some library it uses, for example in [Skulpin](https://github.com/aclysma/skulpin) the buffers & swapchains are essentially created by the Skia bindings with `&dyn HasRawWindowHandle` and [glutin](https://github.com/rust-windowing/glutin/blob/master/glutin_examples/src/lib.rs) can be used to get OpenGL context from `&dyn HasRawWindowHandle` that Skia then [takes in](https://github.com/rust-skia/rust-skia/blob/master/skia-safe/examples/gl-window/main.rs#L113)).

So bringing my miscalleneous thoughts together:
- It'd be nice to control the backend Slint is using pretty comprehensively down to providing your own wayland integrations
- Use cases cover system UI components (overlays, etc) and extra functionality such as pointer locking, gestures, tablet input, drag-and-drop, etc (Only covering cases which cannot be circumvented by just having 2 wayland connections from same app, one for UI and one for business logic)
- Same kind of backend composability has been done in for example [egui](https://github.com/emilk/egui)
- For better developer experience ability to create backend by just providing:
  - something that takes pixel buffer or `&dyn HasRawWindowHandle` and renders to it
  - Ability to provide mouse move, mouse click, keyboard press and etc events from a queue or provided functions. This currently happens in embedded examples already

TBH this looks like being pretty close to the MinimalSoftwareWindow solutions but with ability to replace that by something that can be given `&dyn HasRawWindowHandle` and renders directly to that.

**Comment by HigherOrderLogic** [comment]
Is there any update on this, or are we still waiting for work on Winit's side?

**Reply by ogoffart** [reply]
You can get the raw window handle from Slint, and latest version of Slint allow to access the winit API as well.

Some project have been using the zwlr_layer_shell_v1 api with Slint, for example 
 https://github.com/waydeerwm/layer-shika

**Reply by dotsem** [reply]
I recommend using this project:
[https://github.com/VimYoung/Spell](https://github.com/VimYoung/Spell)
It is still in development but it does work quite well with the layer shell protocol at the moment. It is also better documented than layer-shika.

**Comment by zea64** [comment]
Both layer-shika and Spell seem to smuggle a window list into the Platform and do some finicky management to make sure create_window_adapter is called with components in the right order. Seems like this would be better served by some API to allow instantiating a component with a WindowAdapter directly. I was trying to do it myself, until I saw dotsem mention Spell (I didn't want layer-shika's use of slint-interpreter), and was very much not looking forward to it.

On another note, I noticed that femtovg's GlContextWrapper is private which is a shame since all I *really* needed was a custom window, not custom GL (besides giving it the correct surface).
